validator/testcases/regex.py

```diff
@@ -2732,3 +2736,113 @@ def tests(self):
+            for i, (regexp, data) in enumerate(regexps)))
+
+    def process_key(self, key):
+        """Processes a key into a regular expression. Currently turns an
+        enumerable value into a regexp which matches a string which is
+        exactly equal to any included value."""
+
+        if isinstance(key, basestring):
+            return key
+
+        return r"^(?:%s)$" % "|".join(map(re.escape, key))
+
+    def test(self, string, traverser):
+        for match in self.regex.finditer(string):
+            for key, val in match.groupdict().iteritems():
+                if val is not None and key in self.tests:
+                    kw = {"err_id": ("testcases_regex", "string", "generic")}
+                    kw.update(self.tests[key])
+
+                    traverser.warning(**kw)
+
+
+PROFILE_FILENAMES = (
+    "SiteSecurityServiceState.txt",
+    "addons.json",
+    "addons.sqlite",
+    "blocklist.xml",
+    "cert8.db",
+    "compatibility.ini",
+    "compreg.dat",
+    "content-prefs.sqlite",
+    "cookies.sqlite",
+    "directoryLinks.json",
+    "extensions.ini",
+    "extensions.json",
+    "extensions.sqlite",
+    "formhistory.sqlite",
+    "healthreport/*",
+    "healthreport.sqlite",
+    "httpDataUsage.dat",
+    "key3.db",
+    "localstore.rdf",
+    "logins.json",
+    "permissions.sqlite",
+    "places.sqlite",
+    "places.sqlite-shm",
+    "places.sqlite-wal",
+    "pluginreg.dat",
+    "prefs.js",
+    "user.js",
+    "safebrowsing/*",
+    "search-metadata.json",
+    "search.json",
+    "search.sqlite",
+    "searchplugins/*",
+    "secmod.db",
+    "sessionCheckpoints.json",
+    "sessionstore.js",
+    "signons.sqlite",
+    "startupCache/*",
+    "storage/*",
+    "urlclassifier.pset",
+    "urlclassifier3.sqlite",
+    "urlclassifierkey3.txt",
+    "webapps/*",
+    "webappsstore.sqlite",
+    "xpti.dat",
+    "xulstore.json")
+
+
+def munge_filename(name):
+    # Changes filenames ending with `/*` into a regular expression which
+    # will also match sub-paths across platforms. Escapes regex metacharacters
+    # in other strings.
+    if name.endswith("/*"):
+        return r"%s(?:[/\\].*)?" % re.escape(name[:-2])
+    return re.escape(name)
+
+PROFILE_REGEX = r"(?:^|[/\\])(?:%s)$" % "|".join(map(munge_filename,
+                                                     PROFILE_FILENAMES))
+
+STRING_REGEXPS = (
+    (PROFILE_REGEX, {
+        "err_id": ("testcases_regex", "string", "profile_filenames"),
+        "warning": "Reference to critical user profile data",
+        "description": "Critical files in the user profile should not be "
+                       "directly accessed by add-ons. In many cases, an "
+                       "equivalent API is available and should be used "
+                       "instead.",
+        "signing_severity": "low"}),
+    (DANGEROUS_CATEGORIES, DANGEROUS_CATEGORY_WARNING),
+)
+
+validate_string = RegexTest(STRING_REGEXPS).test
```

Comment:
On Mon, May 18, 2015 at 07:03:13AM -0700, Mathieu Agopian wrote:
>Given the complexity of the code, and the easiness of making mistakes with regular expressions, I think those should be tested (talking about munge_filename, PROFILE_REGEX, validate_string, and RegextTest).

You're probably right. I added tests for all of the permutations 
that these tests should match, but unit tests would definitely 
not hurt. Will do.